File: RaspberryPi/python_demo/11_photo_interrupter/11_photo_interrupter.py
# !/usr/bin/env python3
# -*- coding: utf-8 -*-
# file name: photo_interrupter.py
# version: 1.0
# author: Neal Shen
# ------------------------------------------------------------------
import RPi.GPIO as GPIO
makerobo_LBPin	 	= 11    # photo interrupter pin
makerobo_Rpin   	= 12    # red led pin
makerobo_Gpin   	= 13    # green led pin


# gpio inilization
def makerobo_setup():
    GPIO.setmode(GPIO.BOARD)
    GPIO.setwarnings(False)
    GPIO.setup(makerobo_Rpin, GPIO.OUT)
    GPIO.setup(makerobo_Gpin, GPIO.OUT)
    GPIO.setup(makerobo_LBPin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
    # set BtnPin mode is input, pull up to high(3.3V)
    # interupt function
    GPIO.add_event_detect(makerobo_LBPin, GPIO.BOTH, callback=makerobo_detect, bouncetime=200)


def double_color(x):
    if x == 0:
        GPIO.output(makerobo_Rpin, 1)
        GPIO.output(makerobo_Gpin, 0)
    if x == 1:
        GPIO.output(makerobo_Rpin, 0)
        GPIO.output(makerobo_Gpin, 1)


def makerobo_print(x):
    if x == 0:
        print('******************************************')
        print('* makerobo Light was blocked! *')
        print('******************************************')


# interrupt function
def makerobo_detect(chn):
    double_color(GPIO.input(makerobo_LBPin))
    makerobo_print(GPIO.input(makerobo_LBPin))


# loop function
def makerobo_loop():
    while True:
        pass


# release source
def makerobo_destroy():
    GPIO.output(makerobo_Rpin, GPIO.HIGH)
    GPIO.output(makerobo_Gpin, GPIO.HIGH)
    GPIO.cleanup()


if __name__ == '__main__':
    makerobo_setup()
    try:
        makerobo_loop()
    except KeyboardInterrupt:
        makerobo_destroy()

File: RaspberryPi/cplus_demo/11_photo_interrupter/11_photo_interrupter.c
/**********************************************
 * -----------photo interrupter---------------
 * file name: photo interrupter.c
 * version: 1.0
 * author: Neal Shen
 *********************************************/
#include <wiringPi.h>
#include <stdio.h>

#define makerobo_LBPin			0			// photo interrupter pin
#define makerobo_Rpin			1			// red led pin
#define makerobo_Gpin			2			// green led pin

/**********************************************
 * function name: double_color
 * param: color
 * author: Neal Shen
 *********************************************/
void double_color(int color)
{
	// set Red light pin mode is output
	pinMode(makerobo_Rpin, OUTPUT);
	// set Green light pin mode is output
	pinMode(makerobo_Gpin, OUTPUT);

	if (color == 0)			// enable red led
	{
		digitalWrite(makerobo_Rpin, HIGH);
		digitalWrite(makerobo_Gpin, LOW);
	}
	else if (color == 1)	// enable green led
	{
		digitalWrite(makerobo_Rpin, LOW);
		digitalWrite(makerobo_Gpin, HIGH);
	}
	else
	{
		printf("makerobo double color led error\n");
	}


}

/**********************************************
 * function name: makerobo_Print
 * param: x
 * author: Neal Shen
 *********************************************/
void makerobo_Print(int x)
{
	if (x == 0)
	{
		printf("makerobo Light was blocked\n");
	}
}

int main(void)
{
	int clb_temp;
	// if the initialization is failed, print the infomation
	// on the screen of raspberrypi
	if (wiringPiSetup() == -1)
	{
		printf("setup wiringPi failed !");
		return -1;
	}

	// set photo interrupter pin mode is input
	pinMode(makerobo_LBPin, INPUT);

	while (1) {
		// detect 
		if (0 == digitalRead(makerobo_LBPin))
		{
			clb_temp = 1;
		}
		if (1 == digitalRead(makerobo_LBPin))
		{
			clb_temp = 0;
		}
		double_color(clb_temp);
		makerobo_Print(clb_temp);
	}
	return 0;
}